### PR TITLE
Allow resolving of ES5 default values in functions

### PR DESCRIPTION
**Pull request by alxhub**

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/host/src/reflection.ts [hunk 1]
```diff
@@ -180,6 +180,53 @@ export interface Parameter {
...
+export interface FunctionDefinition<T extends ts.MethodDeclaration|ts.FunctionDeclaration|ts.FunctionExpression> {
+  /**
+   * A reference to the node which declares the function.
+   */
+  node: T;
+
+  /**
+   * Statements of the function body, if a body is present, or null if no body is present.
+   *
+   * This list may be filtered to exclude statements which perform parameter default value
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/host/src/reflection.ts]: Could this be reworded to `This list may have been filtered (...)` to indicate the `ReflectionHost` is expected to do that. Currently it reads to me as if consumers of `ReflectionHost` are responsible for doing so, in which it doesn't make much sense to tell here wat consumers can do with an array :)

File: packages/compiler-cli/src/ngtsc/host/src/reflection.ts [hunk 2]
```diff
@@ -273,7 +320,29 @@ export interface ReflectionHost {
...
+
+  /**
+   * Reflect over a function and return metadata about its parameters and body.
+   *
+   * Functions in TypeScript and ES5 code have different AST representations, in particular around
+   * default values for parameters. A TypeScript function has its default value as the initializer
+   * on the parameter declaration, whereas an ES5 function has its default value set in a statement
+   * of the form:
+   *
+   * if (param === void 0) { param = 3; }
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/host/src/reflection.ts]: Perhaps stick some triple back ticks around this.

File: packages/compiler-cli/src/ngtsc/host/src/reflection.ts [hunk 3]
```diff
@@ -273,7 +320,29 @@ export interface ReflectionHost {
...
+   *
+   * This method abstracts over these details, and interprets the function declaration and body to
+   * extract parameter default values and the "real" body.
+   *
+   * A current limitation is that this metadata has no representation for shorthand assignment of
+   * parameter objects in the function signature.
+   *
+   * @param node a TypeScript `ts.Declaration` node representing the function over which to reflect.
```
Comment (gkalpak) [inline packages/compiler-cli/src/ngtsc/host/src/reflection.ts]: The param's name is `fn` :grin:

File: packages/compiler-cli/src/ngtsc/metadata/src/reflector.ts [hunk 4]
```diff
@@ -141,6 +141,18 @@ export class TypeScriptReflectionHost implements ReflectionHost {
     return this._getDeclarationOfSymbol(symbol);
   }
 
+  getDefinitionOfFunction<T extends ts.FunctionDeclaration|ts.MethodDeclaration|ts.FunctionExpression>(node: T): FunctionDefinition<T> {
+    return {
+      node,
+      body: node.body !== undefined ? Array.from(node.body.statements) : null,
+      parameters: node.parameters.map(node => {
```
Comment (gkalpak) [inline packages/compiler-cli/src/ngtsc/metadata/src/reflector.ts]: Consider renaming the second `node` to `paramNode` to avoid confusion below.

**Review by gkalpak — APPROVED** [review]
LGTM (linting issues aside)

**Comment by alxhub** [comment]
@petebacondarwin will merge this along with the ngcc implementation.

**Pull request closed without merging**